### ISSUE TITLE
feature question: infer a function def for a call

### ISSUE THREAD

**Issue by zhammer** [Instagram/LibCST#86]
I adapted a pylint plugin today and found they had a helpful util that tries to infer the function def for a call node while traversing the AST.

here's how i use `safe_infer`:
```py
def visit_call(self, node):
    if not self.linter.is_message_enabled(self.MESSAGE_ID):
        return

    func_def = utils.safe_infer(node.func)
    # i then check if the called function has a specific coroutine
    ...
```

here's the definition: https://github.com/PyCQA/pylint/blob/47fdef57e6c453ddbd65d4960db20c0e580ee041/pylint/checkers/utils.py#L1077-L1097

is there anything like this being used in libcst tools in use at instagram? it gives a _lot_ more power to linters/codemods imo. i imagine this'd look something like:

1. do an initial first pass to collect metadata function (as well as other) definitions, keeping their file/module path
2. use that metadata in a visitor to help infer the definition of some function/class/etc.

**Comment by jimmylai** [comment]
@zhammer Retrieve the FunctionDef node given a Call is doable through our Scope metadata provider.
https://2589-200896124-gh.circle-artifacts.com/0/doc/metadata.html#libcst.metadata.Scope

You can add the dependency to use the provider in your visitor/transformer
```
class MyVisitor(libcst.CSTVisitor):
    METADATA_DEPENDENCIES = (libcst.metadata.ScopeProvider,)

    def visit_Call(self, node):
        scope = self. get_metadata(libcst.metadata.ScopeProvider, node)
        assignments = scope[node]  # you'll get a list of possible assignments
```
See the Assignment class for more detail.
https://2580-200896124-gh.circle-artifacts.com/0/doc/metadata.html#libcst.metadata.Assignment

I'm working on a couple more powerful metadata providers and plan to add tutorial to demonstrate some use cases.

**Comment by zhammer** [comment]
hey @jimmylai ! any updates on the repository analysis/refactoring? just curious

**Comment by jimmylai** [comment]
@zhammer I'm currently working on it and hoping to deliver it by the end of this month.

**Comment by jimmylai** [comment]
Update: the TypeInferenceProvider PRs are in review and will be available in the next release.
https://github.com/Instagram/LibCST/pull/159
https://github.com/Instagram/LibCST/pull/178
https://github.com/Instagram/LibCST/pull/179

**Comment by jimmylai** [comment]
`TypeInferenceProvider` is available in LibCST 0.2.6. Given a `Call.func`, it provides the type annotation of the node (`Callable(name)[...]`). The name is provided as a fully qualified name. So we can potentially use the name to read source code and parse as CST to get the `FunctionDef` node.
